Exercise7: DataCleaning  

15 Years of Power Outages  
Note: some ideas and suggestions are from Andre Armstrong's Jupyter notebook "How long does an outage last?"  
  


In [134]:
import pandas as pd
import numpy as np
import chardet
import matplotlib.pyplot as plt

In [135]:
# df = pd.read_csv('http://localhost:8888/notebooks/Documents/Martina/Aufgabe7/Grid_Disruption_00_14_standardized.csv')
# must be logged in at kaggle.com to read the data, so we created a local copy
# df = pd.read_csv('https://www.kaggle.com/autunno/15-years-of-power-outages/downloads/Grid_Disruption_00_14_standardized%20-%20Grid_Disruption_00_14_standardized.csv')
# df = pd.read_csv('C:\\Users\\ahlswede\\Documents\\Martina\\Aufgabe7\\Grid_Disruption_00_14_standardized.csv')
# loading from github
df = pd.read_csv('https://raw.githubusercontent.com/mahlswede/DataCleaningExercise/master/Grid_Disruption_00_14_standardized.csv')

In [136]:
df.head(10)
#df.sample(10)

Event Description  Year Date Event Began Time Event Began  \
0  Severe Weather - Thunderstorms  2014        6/30/2014          8:00 PM   
1  Severe Weather - Thunderstorms  2014        6/30/2014         11:20 PM   
2  Severe Weather - Thunderstorms  2014        6/30/2014          5:55 PM   
3    Fuel Supply Emergency - Coal  2014        6/27/2014          1:21 PM   
4     Physical Attack - Vandalism  2014        6/24/2014          2:54 PM   
5     Physical Attack - Vandalism  2014        6/19/2014          8:47 AM   
6     Physical Attack - Vandalism  2014        6/18/2014          9:52 AM   
7  Severe Weather - Thunderstorms  2014        6/18/2014          5:00 PM   
8  Severe Weather - Thunderstorms  2014        6/15/2014         12:00 AM   
9       Suspected Physical Attack  2014        6/12/2014          9:10 AM   

  Date of Restoration Time of Restoration  \
0            7/2/2014             6:30 PM   
1            7/1/2014             5:00 PM   
2            7/1/2014             2:53 AM   
3             Unknown             Unknown   
4           6/24/2014             2:55 PM   
5           6/19/2014             8:48 AM   
6           6/18/2014             7:00 PM   
7           6/20/2014             3:00 PM   
8           6/15/2014             1:00 AM   
9           6/12/2014             9:11 AM   

                                          Respondent         Geographic Areas  \
0                           Exelon Corporation/ComEd                 Illinois   
1            Northern Indiana Public Service Company    North Central Indiana   
2                                        We Energies     Southeast  Wisconsin   
3                                        We Energies                Wisconsin   
4                         Tennessee Valley Authority     Nashville, Tennessee   
5                         Tennessee Valley Authority     Nashville, Tennessee   
6  Public Utility District No.1 of Snohomish Coun...               Washington   
7                                  Detroit Edison Co       Southeast Michigan   
8                                        Xcel Energy        Central Minnesota   
9                       Luminant Energy Company, LLC  Somervell County, Texas   

  NERC Region Demand Loss (MW) Number of Customers Affected  \
0         RFC          Unknown                      420,000   
1         RFC          Unknown                      127,000   
2         MRO              424                      120,000   
3         MRO          Unknown                      Unknown   
4        SERC          Unknown                      Unknown   
5        SERC          Unknown                      Unknown   
6        WECC          Unknown                      Unknown   
7         RFC          Unknown                      138,802   
8         MRO          Unknown                       55,951   
9       ERCOT          Unknown                      Unknown   

                           Tags  
0  severe weather, thunderstorm  
1  severe weather, thunderstorm  
2  severe weather, thunderstorm  
3   fuel supply emergency, coal  
4           vandalism, physical  
5           vandalism, physical  
6           vandalism, physical  
7  severe weather, thunderstorm  
8  severe weather, thunderstorm  
9           vandalism, physical

In [137]:
df.shape

(1652, 12)

our dataset 'GridDisruption' has 1652 records with each 12 columns  
are there missing entries ?

In [138]:
df2 = df.dropna(how='all')
print(df2.shape)
df2 = df.dropna(how='any')
print(df2.shape)

(1652, 12)
(1193, 12)


we don't have records where **all** fields are empty 
but we have 459 records with at least one missing entry  

In [139]:
print(df.isnull().sum())
print("missing entries: " + str(df.isnull().sum().sum()))

Event Description                 0
Year                              0
Date Event Began                  0
Time Event Began                  9
Date of Restoration              14
Time of Restoration              20
Respondent                        0
Geographic Areas                  1
NERC Region                       2
Demand Loss (MW)                398
Number of Customers Affected    214
Tags                              1
dtype: int64
missing entries: 659


overall we have 659 missing values, the table shows the affected columns  
there are a lot of entries with 'Unknown' at fields 'Demand Loss' and 'Number of Customers Affected'  
or entries with 'Ongoing' at fields 'Date..' and 'Time of Restauration'  
we will replace these entries with np.NaN and remove them using dropna()

In [140]:
df.replace('Unknown', np.nan, inplace=True)
df.replace('Ongoing', np.nan, inplace=True)

In [141]:
df.isnull().any()

Event Description               False
Year                            False
Date Event Began                False
Time Event Began                 True
Date of Restoration              True
Time of Restoration              True
Respondent                      False
Geographic Areas                 True
NERC Region                      True
Demand Loss (MW)                 True
Number of Customers Affected     True
Tags                             True
dtype: bool

let's see where are the errors:

In [142]:
print("Total number of rows: " + str(len(df.index)))
print("Number of empty values:")
for column in df.columns:
    print(" * " + column + ": " + str(df[column].isnull().sum()))

Total number of rows: 1652
Number of empty values:
 * Event Description: 0
 * Year: 0
 * Date Event Began: 0
 * Time Event Began: 10
 * Date of Restoration: 34
 * Time of Restoration: 40
 * Respondent: 0
 * Geographic Areas: 9
 * NERC Region: 2
 * Demand Loss (MW): 605
 * Number of Customers Affected: 359
 * Tags: 1


In [143]:
df = df.dropna(how='any')

In [144]:
df.shape

(956, 12)

we lost 696 faulty records, 956 valid records are left:

In [145]:
print("Total number of rows: " + str(len(df.index)))
print("Number of empty values:")
for column in df.columns:
    print(" * " + column + ": " + str(df[column].isnull().sum()))

Total number of rows: 956
Number of empty values:
 * Event Description: 0
 * Year: 0
 * Date Event Began: 0
 * Time Event Began: 0
 * Date of Restoration: 0
 * Time of Restoration: 0
 * Respondent: 0
 * Geographic Areas: 0
 * NERC Region: 0
 * Demand Loss (MW): 0
 * Number of Customers Affected: 0
 * Tags: 0
